### PR TITLE
BHBC-2070: Keycloak Standard Upgrade, DevOps Pipeline Updates, Misc Updates

### PR DESCRIPTION
**Pull request by NickPhura**
# Overview

## Links to Jira tickets

https://quartech.atlassian.net/browse/BHBC-2070

## Description of relevant changes

- Update Keycloak from silver to gold (standard realm)
- Pipeline updates copied from SIMS/Biohub
- Misc updates copied from SIMS/Biohub
- https://youtu.be/1zNdw4DaUM8?t=4

### Update from keycloak silver to gold (copied from Sims/BioHub)
- Test that logging in, access requests, etc, work. IDIR/BCEID_BASIC/BCEID_BUSINESS. Same as from SIMS or BioHub

### Update pipeline and git action files (copied from Sims/BioHub)
- If everything builds correctly in the PR deploy and when merging to Dev.

### Update treatment validation logic (using Zod)
- Will need to get some sample treatment shapefiles 
  - (Kjartan had some old ones, I can edit them a bit to force errors, Robin can also produce them).
- Test that the treatment import works
  - Shows any validation errors to the user
  - If no errors, ingests data correctly and renders in the APP.
  - If subsequent data is ingested, it should be appended to the previous data, I believe.

### Add DB Migration to tweak the treatment_unit table
- New column `implemented` (was in the shapefile data, but never captured in the database for w/e reason)
- Updated column `reconnaissance_conducted` (changed the type, and made optional)

## PR Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

### CODE REVIEW

**Review by curtisupshall — COMMENTED** [review]
Code looks good! Approval pending some manual testing 🤠

**Review by AlfredRosenthal — DISMISSED** [review]
testing this out, uploaded new treatments replacing the existing upload. There is an alert warning the user so I imagine things are working as intended.

🍡

File: database/src/migrations/20230419151100_treatment_unit_updates.ts
Comment (NickPhura) [inline database/src/migrations/20230419151100_treatment_unit_updates.ts:1]: @cgarrettjones FYI: the migration we discussed, with the check constraints, etc.

**Review by AlfredRosenthal — DISMISSED** [review]
🍭

**Review by anissa-agahchen — DISMISSED** [review]

**Review by AlfredRosenthal — APPROVED** [review]
🍬

**Review by anissa-agahchen — APPROVED** [review]

**Pull request merged**